File: Data Structures & Algorithms/find-minimum-in-rotated-sorted-array/submission-4.java
public class Solution {
    public int findMin(int[] nums) {
        int left = 0, right = nums.length - 1, ans = nums[0];

        while (left <= right) {
            if (nums[left] < nums[right]) {
                ans = Math.min(ans, nums[left]);
                break;
            }

            int mid = left + (right - left) / 2;
            ans = Math.min(ans, nums[mid]);
            if (nums[mid] >= nums[left]) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return ans;
    }
}

File: Data Structures & Algorithms/max-water-container/submission-4.cpp
class Solution {
public:
    int maxArea(vector<int>& heights) {
        int left = 0, right = heights.size() - 1, area = 0, maxArea = 0;

        while (left < right) {
            area = (right - left) * min(heights[right], heights[left]);
            maxArea = max(area, maxArea);

            if (heights[left] > heights[right])
                right--;
            else
                left++;
        }

        return maxArea;
    }
};
